src/protocol/boolean/comparison.rs

```diff
@@ -2,10 +2,126 @@ use super::or::or;
 use crate::{
     error::Error,
     ff::Field,
-    protocol::{context::Context, BasicProtocols, BitOpStep, RecordId},
+    protocol::{
+        boolean::{random_bits_generator::RandomBitsGenerator, RandomBits},
+        context::Context,
+        BasicProtocols, BitOpStep, RecordId,
+    },
     secret_sharing::Arithmetic as ArithmeticSecretSharing,
 };
 
+// Adapted from 6.1 Interval Test Protocol in "Multiparty Computation for Interval, Equality, and
+// Comparison Without Bit-Decomposition Protocol", Nishide & Ohta, PKC 2007.
+// <https://doi.org/10.1007/978-3-540-71677-8_23>
+//
+// The version in the paper tests c_1 < x < c_2. For us, c_1 is zero (which eliminates the `c < c_1`
+// case enumerated in the paper), we test ≤ rather than <, and we finally return `x > c_2` which is
+// ~(0 ≤ x ≤ c_2).
+//
+// The remainder of this description names the variables consistently with other routines in this
+// file, which map to the paper as follows:
+//
+// Paper  Ours
+// x      a
+// c      b
+// c_1    0
+// c_2    c
+//
+// Goal: Compute `a > c` as `~(0 ≤ a ≤ c)`
+//
+// Strategy:
+//  1. Generate random r
+//  2. Reveal b = a + r
+//  3. Derive bounds r_low and r_high from public values, such that the desired result is a simple
+//     function of `r_low < r` and `r < r_high`.
+//
+// Case 1: b > c
+// b - c - 1 < r < b + 1
+//  ⟺  b - c ≤ r ≤ b
+//  ⟺  b - b ≤ b - r ≤ b - (b - c)
+//  ⟺  0 ≤ a ≤ c
+//  ⟺  ~(a > c)
+//
+// Case 2: b ≤ c
+//  b < r < b + p - c
+//  ⟺  b - (b + p - c) < b - r < b - b
+//  ⟺  -(p - c) < a < 0
+//  ⟺  a > c
```

Comment:
This is brilliant. Great work finding this excellent algorithm @andyleiserson! But wow... there are a lot of opportunities for an "off-by-one" error here. I tried to work the math out on paper myself, and all of the < vs <= signs, and the extra +1 terms here and there are *tricky* to get right.

I'm sure your implementation is correct, but I agree with @martinthomson it would be great to see some tests for this `greater_than_constant` function.

Man, I can't wait to merge this though!

Reply:
it would be great to add some property tests here, to make sure we cover larger inputs. Other than that, lgtm. Great work!

Reply:
I didn't see how to write a proptest for async code, but with some tweaks and a small amount of duplication, I was able to proptest a non-async version.